Neos.Flow/Classes/Mvc/Routing/Dto/UriConstraints.php

```diff
@@ -0,0 +1,199 @@
+<?php
+namespace Neos\Flow\Mvc\Routing\Dto;
+
+/*
+ * This file is part of the Neos.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Http\Uri;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * @Flow\Proxy(false)
+ */
+final class UriConstraints
+{
+    const CONSTRAINT_SCHEME = 'scheme';
+    const CONSTRAINT_HOST = 'host';
+    const CONSTRAINT_SUB_DOMAIN = 'subDomain';
+    const CONSTRAINT_TOP_LEVEL_DOMAIN = 'topLevelDomain';
+    const CONSTRAINT_PORT = 'port';
+    const CONSTRAINT_PATH = 'path';
+    const CONSTRAINT_PATH_PREFIX = 'pathPrefix';
+    const CONSTRAINT_PATH_SUFFIX = 'pathSuffix';
+
+    /**
+     * @var array
+     */
+    private $constraints;
+
+    private function __construct(array $constraints)
+    {
+        $this->constraints = $constraints;
+    }
+
+    public static function create(): self
+    {
+        return new static([]);
+    }
+
+    public function merge(UriConstraints $uriConstraints): self
+    {
+        $mergedConstraints = array_merge($this->constraints, $uriConstraints->constraints);
+        return new static($mergedConstraints);
+    }
+
+    public function withScheme(string $scheme): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_SCHEME] = $scheme;
+        return new static($newConstraints);
+    }
+
+    public function withHost(string $host): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_HOST] = $host;
+        return new static($newConstraints);
+    }
+
+    public function withSubDomain(string $subDomain): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_SUB_DOMAIN] = $subDomain;
+        return new static($newConstraints);
+    }
+
+    public function withTopLevelDomain(string $topLevelDomain): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_TOP_LEVEL_DOMAIN] = $topLevelDomain;
+        return new static($newConstraints);
+    }
+
+    public function withPort(int $port): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_PORT] = $port;
+        return new static($newConstraints);
+    }
+
+    public function withPath(string $path): self
+    {
+        $newConstraints = $this->constraints;
+        $newConstraints[self::CONSTRAINT_PATH] = $path;
+        return new static($newConstraints);
+    }
+
+    public function withPathPrefix(string $pathPrefix): self
+    {
+        if ($pathPrefix === '') {
+            return $this;
```

Comment:
Good catch because `withPathPrefix` always *adds* the prefix and won't replace existing ones. But I don't think it would be necessary to *remove* any previously added prefixes.
The way this works currently is that the (Noes Frontend) Route Part adds a domain specific prefix like `de/` and afterwards the Router prefixes `index.php/` if URL rewriting is not available. I could imagine other Route Parts adding more prefixes but they should not unset/overwrite previous ones as that would be highly confusing.. Anyways, if the need arises we can always add this